src/SocketTrebuchet.ts

```diff
@@ -56,7 +99,43 @@ class SocketTrebuchet extends Trebuchet {
         this.keepAlive()
         this.ws.send(PONG)
       } else {
-        this.emit('data', this.decode(data))
+        const {message, mid} = this.decode(data)
+        if (!message) return
+        // handle non-reliable messages
+        if (typeof mid !== 'number' || mid > MAX_MID || mid < 0) {
+          this.emit('data', message)
+          return
+        }
+
+        // ignore mids that were already requested & processed
+        if (this.midsToIgnore.includes(mid)) return
+
+        // TCP guarantees AT MOST ONCE
+        // if a mid is requested, then the delivery guarantee becomes AT LEAST ONCE, so we need to ignore dupes
+        if (this.requestedMids.includes(mid)) {
+          this.midsToIgnore.push(mid)
+          if (this.midsToIgnore.length > MAX_IGNORE_LEN) {
+            this.midsToIgnore.splice(0, 1)
+          }
+        }
+
+        this.sendAck(mid)
+
+        if (this.lastMid + 1 === mid) {
+          // get next in order
+          this.lastMid = mid
+          this.emit('data', message)
+          this.releaseNextRobustMessage()
+          return
+        }
+
+        // a missing message was detected!
+        this.robustQueue[mid] = message
+
+        // request the first missing message (could get called multiple times)
+        const missingMid = this.lastMid + 1
+        this.sendReq(missingMid)
```

Comment:
if message 5 comes in & the last one we got was 2, it will re-request message 3.
this is fine, because it makes no sense to request message 4 before we get message 3.